subsys/bluetooth/host/gatt.c

```diff
@@ -2171,16 +2186,58 @@ bool bt_gatt_is_subscribed(struct bt_conn *conn,
 }
 
 #if defined(CONFIG_BT_GATT_CLIENT)
+static struct gatt_sub *gatt_sub_find(struct bt_conn *conn)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(subscriptions); i++) {
+		struct gatt_sub *sub = &subscriptions[i];
+
+		if (conn) {
+			if (!bt_conn_addr_le_cmp(conn, &sub->peer)) {
+				return sub;
+			}
+		} else if (!bt_addr_le_cmp(BT_ADDR_LE_ANY, &sub->peer)) {
+			return sub;
+		}
+	}
+
+	return NULL;
+}
+
+static struct gatt_sub *gatt_sub_add(struct bt_conn *conn)
+{
+	struct gatt_sub *sub;
+
+	sub = gatt_sub_find(conn);
+	if (sub) {
+		return sub;
+	}
+
+	sub = gatt_sub_find(NULL);
+	if (sub) {
+		bt_addr_le_copy(&sub->peer, &conn->le.dst);
+		return sub;
+	}
```

Comment:
The pattern I've used elsewhere for this, in order to avoid iterating the list twice, is to pass a second "free_slot" parameter to the lookup function. The lookup function will set that to point to the first empty slot in case the lookup of the actual connection fails: https://github.com/zephyrproject-rtos/zephyr/blob/8067aca7e90617fb49fb0895066c164d60559198/subsys/bluetooth/mesh/cfg_srv.c#L1107-L1131

Would something like that make sense here?

Reply:
To be honest I haven't consider this would be a big deal since MAX_CONN+MAX_PAIRED is normally not that big, but I can use the same logic to just return an unused slot in one go.

Reply:
Btw an alternative for this is to use ffs and then maintain a bitmap i.e:

```
/* Find unique id in range from 1 to max but no bigger then
 * sizeof(int) * 8. ffs() is used since it is POSIX standard
 */
uint8_t util_get_uid(unsigned int *bitmap, uint8_t max)
{
	uint8_t id;

	id = ffs(~*bitmap);

	if (!id || id > max)
		return 0;

	*bitmap |= 1 << (id - 1);

	return id;
}

/* Clear id bit in bitmap */
void util_clear_uid(unsigned int *bitmap, uint8_t id)
{
	if (!id)
		return;

	*bitmap &= ~(1 << (id - 1));
}
```